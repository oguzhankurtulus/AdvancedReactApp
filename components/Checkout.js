import { CardElement, Elements, useElements, useStripe } from '@stripe/react-stripe-js';
import { loadStripe } from '@stripe/stripe-js';
import nProgress from 'nprogress';
import { useState } from 'react';
import styled from 'styled-components';
import SickButton from './styles/SickButton';

const CheckoutFormStyles = styled.form`
    box-shadow: 0 1px 2px 2px rgba(0, 0, 0, 0.04);
    border: 1px solid rgba(0, 0, 0, 0.06);
    border-radius: 5px;
    padding: 1rem;
    display: grid;
    grid-gap: 1rem;
`;

const stripeLib = loadStripe(process.env.NEXT_PUBLIC_STRIPE_KEY);

export default function CheckoutForm() {
    const [error, setError] = useState(false);
    const [loading, setLoading] = useState(false);
    const stripe = useStripe();
    const elements = useElements();

    const handleSubmit = async (e) => {
        e.preventDefault();
        setLoading(true);

        nProgress.start();

        const { error, paymentMethod } = await stripe.createPaymentMethod({
            type: 'card',
            card: elements.getElement(CardElement),
        });

        if (error) {
            setError(error);
        }

        setLoading(false);
        nProgress.done();
    };
    return (
        <CheckoutFormStyles onSubmit={handleSubmit}>
            {error && <p style={{ fontSize: 12 }}>{error.message}</p>}
            <CardElement />
            <SickButton>Check Out Now</SickButton>
        </CheckoutFormStyles>
    );
}

function Checkout() {
    return (
        <Elements stripe={stripeLib}>
            <CheckoutForm />
        </Elements>
    );
}

export { Checkout };
