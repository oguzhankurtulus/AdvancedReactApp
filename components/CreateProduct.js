import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import Router from 'next/router';
import useForm from '../lib/useForm';
import Form from './styles/Form';
import DisplayError from './ErrorMessage';
import { ALL_PRODUCTS_QUERY } from './Products';

const CREATE_PRODUCT_MUTATION = gql`
    mutation CREATE_PRODUCT_MUTATION(
        # Which variables are getting passed in? And What types are they
        $name: String!
        $description: String!
        $price: Int!
        $image: Upload
    ) {
        createProduct(
            data: {
                name: $name
                description: $description
                price: $price
                status: "AVAILABLE"
                photo: { create: { image: $image, altText: $name } }
            }
        ) {
            id
            price
            description
            name
        }
    }
`;

function CreateProduct() {
    const { inputs, handleChange, clearForm } = useForm({
        image: '',
        name: 'Nike Shoes',
        price: 123,
        description: 'Comfortable Shoes',
    });

    const [createProduct, { loading, error }] = useMutation(CREATE_PRODUCT_MUTATION, {
        variables: inputs,
        refetchQueries: [{ query: ALL_PRODUCTS_QUERY }],
    });

    const handleSubmit = async (e) => {
        e.preventDefault();

        // Submit the input field to the backend
        const res = await createProduct();
        clearForm();
        // Go to that product's page!

        Router.push({
            pathname: `/product/${res.data.createProduct.id}`,
        });
    };
    return (
        <Form onSubmit={handleSubmit}>
            <DisplayError error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
                <label htmlFor="image">
                    Image
                    <input required type="file" id="image" name="image" onChange={handleChange} />
                </label>
                <label htmlFor="name">
                    Name
                    <input
                        type="text"
                        id="name"
                        name="name"
                        placeholder="name"
                        value={inputs.name}
                        onChange={handleChange}
                    />
                </label>
                <label htmlFor="price">
                    Price
                    <input
                        type="number"
                        id="price"
                        name="price"
                        placeholder="price"
                        value={inputs.price}
                        onChange={handleChange}
                    />
                </label>

                <label htmlFor="description">
                    Description
                    <textarea
                        id="description"
                        name="description"
                        placeholder="description"
                        value={inputs.description}
                        onChange={handleChange}
                    />
                </label>
                <button type="submit">+ Add Product</button>
            </fieldset>
        </Form>
    );
}

export default CreateProduct;
