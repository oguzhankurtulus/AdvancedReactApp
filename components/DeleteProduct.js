import { gql } from 'graphql-tag';
import { useMutation } from '@apollo/client';

const DELETE_PRODUCT_MUTATION = gql`
    mutation DELETE_PRODUCT_MUTATION($id: ID!) {
        deleteProduct(id: $id) {
            id
            name
        }
    }
`;

// Remove deleted item from cache without network query
function update(cache, payload) {
    cache.evict(cache.identify(payload.data.deleteProduct));
}

function DeleteProduct({ id, children }) {
    const [deleteProduct, { loading }] = useMutation(DELETE_PRODUCT_MUTATION, {
        variables: {
            id,
        },
        update,
    });
    return (
        <button
            type="button"
            disabled={loading}
            onClick={() => {
                if (confirm('Are you sure you want to delete this item')) {
                    deleteProduct(id).catch((err) => alert(err.message));
                }
            }}
        >
            {children}
        </button>
    );
}

export default DeleteProduct;
