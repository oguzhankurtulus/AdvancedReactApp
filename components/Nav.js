import Link from 'next/link';

function Nav() {
    return (
        <nav>
            <Link href="/products">Products</Link>
            <Link href="/sell">Sell</Link>
            <Link href="/orders">Orders</Link>
            <Link href="/products">Account</Link>
        </nav>
    );
}

export default Nav;
