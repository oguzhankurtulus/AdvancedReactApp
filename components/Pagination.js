import Head from 'next/head';
import Link from 'next/link';
import gql from 'graphql-tag';
import { useQuery } from '@apollo/client';
import PaginationStyles from './styles/PaginationStyles';
import DisplayError from './ErrorMessage';
import { perPage } from '../config';

const PAGINATION_QUERY = gql`
    query {
        _allProductsMeta {
            count
        }
    }
`;

function Pagination({ page }) {
    const { error, loading, data } = useQuery(PAGINATION_QUERY);
    if (loading) {
        return 'Loading...';
    }
    if (error) {
        return <DisplayError error={error} />;
    }
    const { count } = data._allProductsMeta;
    const pageCount = Math.ceil(count / perPage);
    return (
        <PaginationStyles>
            <Head>
                <title>Sick Fits - Page {page} _------</title>
            </Head>
            <Link href={`/products/${page - 1}`}>
                <a aria-disabled={page <= 1}>Prev</a>
            </Link>
            <p>
                Page {page} of {pageCount}
            </p>
            <p> {count} Items Total</p>
            <Link href={`/products/${page + 1}`}>
                <a aria-disabled={page >= 1}>Next</a>
            </Link>
        </PaginationStyles>
    );
}

export default Pagination;
