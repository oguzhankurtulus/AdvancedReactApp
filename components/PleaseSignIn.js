import { useUser } from './User';
import Signin from './Signin';

export default function PleaseSignIn({ children }) {
    const me = useUser();
    if (!me) return <Signin />;
    return children;
}
