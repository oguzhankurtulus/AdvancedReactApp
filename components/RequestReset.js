import { gql, useMutation } from '@apollo/client';
import useForm from '../lib/useForm';
import Form from './styles/Form';
// import { CURRENT_USER_QUERY } from './User';
import Error from './ErrorMessage';

export const REQUEST_RESET_MUTATION = gql`
    mutation REQUEST_RESET_MUTATION($email: String!) {
        sendUserPasswordResetLink(email: $email) {
            code
            message
        }
    }
`;

function RequestReset() {
    const { inputs, handleChange, resetForm } = useForm({
        email: '',
    });

    const [signup, { data, error, loading }] = useMutation(REQUEST_RESET_MUTATION, {
        variables: inputs,
        // refetchQueries: [{ query: CURRENT_USER_QUERY }],
    });

    const handleSubmit = async (e) => {
        e.preventDefault();
        await signup();
        resetForm();
    };
    return (
        <Form method="POST" onSubmit={handleSubmit}>
            <h2>Request a Password Reset</h2>
            <Error error={error} />
            <fieldset disabled={loading}>
                {data?.sendUserPasswordResetLink === null && <p>Success! Check your email for a link!</p>}
                <label htmlFor="email">
                    Email
                    <input
                        type="email"
                        name="email"
                        autoComplete="email"
                        placeholder="Your email address"
                        value={inputs.email}
                        onChange={handleChange}
                    />
                </label>
                <button type="submit" data-testid="resetButton">
                    Request Reset
                </button>
            </fieldset>
        </Form>
    );
}

export default RequestReset;
