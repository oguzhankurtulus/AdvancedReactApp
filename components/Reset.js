import { gql, useMutation } from '@apollo/client';
import Form from './styles/Form';
import useForm from '../lib/useForm';
// import { CURRENT_USER_QUERY } from './User';
import Error from './ErrorMessage';

const RESET_MUTATION = gql`
    mutation RESET_MUTATION($email: String!, $token: String!, $password: String!) {
        redeemUserPasswordResetToken(email: $email, token: $token, password: $password) {
            code
            message
        }
    }
`;

function Reset({ token }) {
    const { inputs, handleChange, resetForm } = useForm({
        email: '',
        password: '',
        token,
    });

    const [reset, { data, loading, error }] = useMutation(RESET_MUTATION, {
        variables: inputs,
    });

    const handleSubmit = async (e) => {
        e.preventDefault();
        const res = await reset().catch(console.error);
        console.log(res);
        resetForm();
    };

    const errorData = data?.redeemUserPasswordResetToken?.code ? data?.redeemUserPasswordResetToken : undefined;

    return (
        <Form method="POST" onSubmit={handleSubmit}>
            <h2>Reset Your Password</h2>
            <Error error={errorData || error} />
            <fieldset disabled={loading}>
                {data?.redeemUserPasswordResetToken === null && <p>Success! You can now sign in!</p>}
                <label htmlFor="email">
                    Email
                    <input
                        type="email"
                        name="email"
                        autoComplete="email"
                        placeholder="Your email address"
                        value={inputs.email}
                        onChange={handleChange}
                    />
                </label>
                <label htmlFor="password">
                    New Password
                    <input
                        type="password"
                        name="password"
                        autoComplete="password"
                        placeholder="New password"
                        value={inputs.password}
                        onChange={handleChange}
                    />
                </label>
                <button type="submit">Request Reset</button>
            </fieldset>
        </Form>
    );
}

export default Reset;
