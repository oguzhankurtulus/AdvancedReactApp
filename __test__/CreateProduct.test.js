import { MockedProvider } from '@apollo/react-testing';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Router from 'next/router'; // Mock this
import wait from 'waait';
import CreateProduct, { CREATE_PRODUCT_MUTATION } from '../components/CreateProduct';
import { ALL_PRODUCTS_QUERY } from '../components/Products';
import { fakeItem } from '../lib/testUtils';

const item = fakeItem();

jest.mock('next/router', () => ({
    push: jest.fn(),
}));

describe('<CreateProduct />', () => {
    it('renders and matches snapshot', () => {
        const { container } = render(
            <MockedProvider>
                <CreateProduct />
            </MockedProvider>
        );
        expect(container).toMatchSnapshot();
    });

    it('handles the updating', async () => {
        render(
            <MockedProvider>
                <CreateProduct />
            </MockedProvider>
        );

        await userEvent.type(screen.getByPlaceholderText(/name/i), item.name);
        await userEvent.type(screen.getByPlaceholderText(/price/i), item.price.toString());
        await userEvent.type(screen.getByPlaceholderText(/description/i), item.description);

        expect(screen.getByDisplayValue(item.name)).toBeInTheDocument();
        expect(screen.getByDisplayValue(item.description)).toBeInTheDocument();
    });

    it('creates the items when the form is submitted', async () => {
        const mocks = [
            {
                request: {
                    query: CREATE_PRODUCT_MUTATION,
                    variables: {
                        name: item.name,
                        description: item.description,
                        image: '',
                        price: item.price,
                    },
                },
                result: {
                    data: {
                        createProduct: {
                            ...item,
                            id: 'abc123',
                            __typename: 'Item',
                        },
                    },
                },
            },
            {
                request: { query: ALL_PRODUCTS_QUERY, variables: { skip: 0, first: 2 } },
                result: {
                    data: {
                        allProducts: [item],
                    },
                },
            },
        ];

        render(
            <MockedProvider mocks={mocks}>
                <CreateProduct />
            </MockedProvider>
        );

        await userEvent.type(screen.getByPlaceholderText(/name/i), item.name);
        await userEvent.type(screen.getByPlaceholderText(/price/i), item.price.toString());
        await userEvent.type(screen.getByPlaceholderText(/description/i), item.description);

        await userEvent.click(screen.getByTestId('submitButton'));
        await waitFor(() => wait(10));
        expect(Router.push).toHaveBeenCalled();
        expect(Router.push).toHaveBeenCalledWith({ pathname: `/product/${item.id}` });
    });
});
