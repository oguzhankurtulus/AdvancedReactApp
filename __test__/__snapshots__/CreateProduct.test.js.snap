// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateProduct /> renders and matches snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 dbjMUV"
  >
    <fieldset
      aria-busy="false"
    >
      <label
        for="image"
      >
        Image
        <input
          id="image"
          name="image"
          required=""
          type="file"
        />
      </label>
      <label
        for="name"
      >
        Name
        <input
          id="name"
          name="name"
          placeholder="name"
          type="text"
          value=""
        />
      </label>
      <label
        for="price"
      >
        Price
        <input
          id="price"
          name="price"
          placeholder="price"
          type="number"
          value="0"
        />
      </label>
      <label
        for="description"
      >
        Description
        <textarea
          id="description"
          name="description"
          placeholder="description"
        />
      </label>
      <button
        data-testid="submitButton"
        type="submit"
      >
        + Add Product
      </button>
    </fieldset>
  </form>
</div>
`;
